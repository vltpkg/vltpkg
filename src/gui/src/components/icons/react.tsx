export const React = (props: React.SVGProps<SVGSVGElement>) => {
  return (
    <svg
      width="500"
      height="500"
      viewBox="0 0 500 500"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
      {...props}>
      <g clip-path="url(#clip0_2967_809)">
        <path
          d="m250.504 284.862c270.046 284.862 285.889 269.029 285.889 249.499c285.889 229.968 270.046 214.136 250.504 214.136c230.961 214.136 215.119 229.968 215.119 249.499c215.119 269.029 230.961 284.862 250.504 284.862z"
          fill="currentColor"
        />
        <path
          d="m250.501 321.96c355.363 321.96 440.371 289.522 440.371 249.509c440.371 209.495 355.363 177.058 250.501 177.058c145.639 177.058 60.6318 209.495 60.6318 249.509c60.6318 289.522 145.639 321.96 250.501 321.96z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.724 285.725c240.155 376.482 310.768 433.837 345.442 413.83c380.116 393.823 365.721 304.031 313.29 213.274c260.859 122.517 190.247 65.1621 155.572 85.1689c120.898 105.176 135.293 194.968 187.724 285.725z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.726 213.277c135.295 304.034 120.9 393.826 155.574 413.833c190.248 433.84 260.861 376.485 313.292 285.728c365.723 194.971 380.118 105.179 345.443 85.172c310.769 65.1652 240.157 122.52 187.726 213.277z"
          stroke="currentColor"
          stroke-width="18.913"
        />
      </g>
      <defs>
        <clipPath id="clip0_2967_809">
          <rect
            width="397"
            height="353"
            fill="currentColor"
            transform="translate(52 73)"
          />
        </clipPath>
      </defs>
    </svg>
  )
}
