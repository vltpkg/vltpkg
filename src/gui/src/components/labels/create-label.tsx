import { useEffect, useState } from 'react'
import { Button } from '@/components/ui/button.jsx'
import { LabelBadge } from '@/components/labels/label-badge.jsx'
import { Label } from '@/components/ui/label.jsx'
import { Input } from '@/components/ui/input.jsx'
import {
  Popover,
  PopoverContent,
  PopoverTrigger,
} from '@/components/ui/popover.jsx'
import { ColorPicker } from '@/components/ui/color-picker.jsx'
import { Palette } from 'lucide-react'
import { useGraphStore } from '@/state/index.js'
import { type Color } from '@/state/types.js'
import { useToast } from '@/components/hooks/use-toast.js'
import { v4 as uuidv4 } from 'uuid'

interface CreateLabelProps {
  closeCreate: () => void
}

const CreateLabel = ({ closeCreate }: CreateLabelProps) => {
  const [labelName, setLabelName] = useState<string>('Label preview')
  const [labelDescription, setLabelDescription] = useState<string>('')
  const [selectedColor, setSelectedColor] = useState<Color>('#00FF5F')
  const [isCreationValid, setIsCreationValid] =
    useState<boolean>(false)
  const saveQueryLabel = useGraphStore(state => state.saveQueryLabel)
  const { toast } = useToast()

  const handleCancel = () => {
    closeCreate()
  }

  const handleCreate = () => {
    saveQueryLabel({
      id: uuidv4(),
      name: labelName,
      description: labelDescription,
      color: selectedColor,
    })
    toast({
      title: `${labelName} saved successfully`,
    })
    closeCreate()
  }

  useEffect(() => {
    if (
      labelName.trim() !== '' &&
      labelName.trim() !== 'Label preview'
    ) {
      setIsCreationValid(true)
    } else {
      setIsCreationValid(false)
    }
  }, [labelName, labelDescription, selectedColor])

  return (
    <div className="rounded-sm border border-[1px] border-muted-foreground/25 bg-neutral-100 transition-all dark:bg-neutral-950">
      <div className="flex flex-col gap-3 px-3 py-3">
        {/* label preview */}
        <div>
          <LabelBadge
            name={
              labelName.trim() !== '' ? labelName : 'Label preview'
            }
            color={selectedColor}
          />
        </div>
        <div className="mt-4 flex w-full gap-3">
          <div className="flex flex-col gap-2">
            <Label className="border-none text-sm font-medium">
              Name
            </Label>
            <Input
              type="text"
              value={labelName}
              onChange={e => setLabelName(e.target.value)}
              placeholder="Name"
            />
          </div>
          <div className="flex grow flex-col gap-2">
            <Label className="border-none text-sm font-medium">
              Description
            </Label>
            <Input
              type="text"
              value={labelDescription}
              onChange={e => setLabelDescription(e.target.value)}
              placeholder="Description (optional)"
            />
          </div>
          <div className="flex flex-col gap-2">
            <Label className="border-none text-sm font-medium">
              Color
            </Label>
            <Popover>
              <PopoverTrigger asChild>
                <Button className="w-[120px]" variant="outline">
                  <Palette />
                  <span>{selectedColor}</span>
                </Button>
              </PopoverTrigger>
              <PopoverContent>
                <ColorPicker
                  defaultColor={selectedColor}
                  defaultInput={selectedColor}
                  onChange={setSelectedColor}
                />
              </PopoverContent>
            </Popover>
          </div>
          <div className="flex items-end justify-end gap-3">
            <Button onClick={handleCancel} variant="outline">
              Cancel
            </Button>
            <Button
              disabled={!isCreationValid}
              onClick={handleCreate}>
              Create label
            </Button>
          </div>
        </div>
      </div>
    </div>
  )
}

export { CreateLabel }
