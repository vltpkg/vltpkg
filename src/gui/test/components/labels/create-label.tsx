import React from 'react'
import { vi, describe, it, expect, afterEach } from 'vitest'
import { cleanup, render } from '@testing-library/react'
import html from 'diffable-html'
import { useGraphStore as useStore } from '@/state/index.js'
import { CreateLabel } from '@/components/labels/create-label.jsx'

vi.mock('@/components/ui/button.jsx', () => ({
  Button: 'gui-button',
}))

vi.mock('@/components/labels/label-badge.jsx', () => ({
  LabelBadge: 'gui-label-badge',
}))

vi.mock('@/components/ui/label.jsx', () => ({
  Label: 'gui-label',
}))

vi.mock('@/components/ui/input.jsx', () => ({
  Input: 'gui-input',
}))

vi.mock('@/components/ui/popover.jsx', () => ({
  Popover: 'gui-popover',
  PopoverContent: 'gui-popover-content',
  PopoverTrigger: 'gui-popover-trigger',
}))

vi.mock('@/components/ui/color-picker.jsx', () => ({
  ColorPicker: 'gui-color-picker',
}))

vi.mock('lucide-react', () => ({
  Palette: 'gui-palette-icon',
}))

expect.addSnapshotSerializer({
  serialize: v => html(v),
  test: () => true,
})

afterEach(() => {
  const CleanUp = () => (useStore(state => state.reset)(), '')
  render(<CleanUp />)
  cleanup()
})

describe('create-label', () => {
  const setStateMock = vi.fn()
  vi.spyOn(React, 'useState').mockReturnValue(['', setStateMock])

  it('should render correctly', () => {
    const Container = () => {
      return <CreateLabel closeCreate={setStateMock} />
    }

    const { container } = render(<Container />)

    expect(container.innerHTML).toMatchSnapshot()
  })
})
