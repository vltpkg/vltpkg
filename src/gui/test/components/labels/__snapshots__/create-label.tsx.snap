// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`create-label > should render correctly 1`] = `

<div class="rounded-sm border border-[1px] border-muted-foreground/25 bg-neutral-100 transition-all dark:bg-neutral-950">
  <div class="flex flex-col gap-3 px-3 py-3">
    <div>
      <gui-label-badge
        name="Label preview"
        color="#00FF5F"
      >
      </gui-label-badge>
    </div>
    <div class="mt-4 flex w-full gap-3">
      <div class="flex flex-col gap-2">
        <gui-label classname="border-none text-sm font-medium">
          Name
        </gui-label>
        <gui-input
          type="text"
          value="Label preview"
          placeholder="Name"
        >
        </gui-input>
      </div>
      <div class="flex grow flex-col gap-2">
        <gui-label classname="border-none text-sm font-medium">
          Description
        </gui-label>
        <gui-input
          type="text"
          value
          placeholder="Description (optional)"
        >
        </gui-input>
      </div>
      <div class="flex flex-col gap-2">
        <gui-label classname="border-none text-sm font-medium">
          Color
        </gui-label>
        <gui-popover>
          <gui-popover-trigger aschild="true">
            <gui-button
              classname="w-[120px]"
              variant="outline"
            >
              <gui-palette-icon>
              </gui-palette-icon>
              <span>
                #00FF5F
              </span>
            </gui-button>
          </gui-popover-trigger>
          <gui-popover-content>
            <gui-color-picker
              defaultcolor="#00FF5F"
              defaultinput="#00FF5F"
            >
            </gui-color-picker>
          </gui-popover-content>
        </gui-popover>
      </div>
      <div class="flex items-end justify-end gap-3">
        <gui-button variant="outline">
          Cancel
        </gui-button>
        <gui-button disabled="true">
          Create label
        </gui-button>
      </div>
    </div>
  </div>
</div>

`;
